### PR TITLE
feat(chartmuseum): Add unit tests for chartmuseum clients

### PR DESCRIPTION
**Pull request by tompizmor**
This PR adds a unit test for the SyncPendingChart function using the real syncer object (not fake one). For the moment, both source and target repo are chartmuseum kinds.

A fake implementation of chartmuseum has been also added to handle all requests from source and target repos. This implementation should be easy to extend and add support for the rest of backends. (I will do that on subsequent PRs)

### CODE REVIEW

File: pkg/client/chartmuseum/chartmuseumtester.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	if got, want := username, rt.username; got != want {
+		rt.t.Errorf("got: %q, want: %q", got, want)
+	}
+	if got, want := password, rt.password; got != want {
+		rt.t.Errorf("got: %q, want: %q", got, want)
+	}
+
+	// Handle recognized requests.
+	if base, chart := path.Split(r.URL.Path); base == "/api/charts/" && r.Method == "GET" {
+		rt.ChartGet(w, r, chart)
```
Comment (jotadrilo) [inline pkg/client/chartmuseum/chartmuseumtester.go:102]: please name it the other way around: `GetChart`

File: pkg/client/chartmuseum/chartmuseumtester.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+	// Whether the repo should load an empty index or not
+	emptyIndex bool
+
+	// index.yaml to be loaded for testing purposes
+	indexFile string
+	// Set to simulate HTTP error responses for specific API calls.
+	ChartsPostError *httpError
+}
+
+//NewTester start a fake HTTP server to handle requests and return a RepoTester object with useful info for testing
```
Comment (jotadrilo) [inline pkg/client/chartmuseum/chartmuseumtester.go:67]: check indentation and format

File: pkg/syncer/sync_test.go [hunk 3]
```diff
@@ -58,3 +65,117 @@ func TestSyncPendingCharts(t *testing.T) {
...
+			}
+			tTester, tCleanup, err := core.NewClientV2Tester(t, tc.targetRepo.GetRepo(), true, "")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer sCleanup()
+			defer tCleanup()
+
+			// Replace URL with source url
+			read, err := ioutil.ReadFile(dstIndex)
```
Comment (jotadrilo) [inline pkg/syncer/sync_test.go:128]: let's use a better name for `read`

File: pkg/syncer/sync_test.go [hunk 4]
```diff
@@ -58,3 +65,117 @@ func TestSyncPendingCharts(t *testing.T) {
...
+				t.Fatal(err)
+			}
+			tTester, tCleanup, err := core.NewClientV2Tester(t, tc.targetRepo.GetRepo(), true, "")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer sCleanup()
+			defer tCleanup()
+
+			// Replace URL with source url
```
Comment (jotadrilo) [inline pkg/syncer/sync_test.go:127]:
```suggestion
			// Replace placeholder URL with source url
```

File: testdata/etcd-index.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    - function
+    - bitnami
+    maintainers:
+    - email: containers@bitnami.com
+      name: Bitnami
+    name: common
+    sources:
+    - https://github.com/bitnami/charts
+    urls:
+    - https://fake.chart.repo.com/testing/common-0.2.1.tgz
```
Comment (jotadrilo) [inline testdata/etcd-index.yaml:25]:
```suggestion
    - TEST_PLACEHOLDER/common-0.2.1.tgz
```

File: testdata/etcd-index.yaml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    - cache
+    - key-value
+    maintainers:
+    - email: containers@bitnami.com
+      name: Bitnami
+    name: etcd
+    sources:
+    - https://github.com/bitnami/bitnami-docker-etcd
+    urls:
+    - https://fake.chart.repo.com/testing/etcd-4.8.0.tgz
```
Comment (jotadrilo) [inline testdata/etcd-index.yaml:50]:
```suggestion
    - TEST_PLACEHOLDER/etcd-4.8.0.tgz
```

File: pkg/syncer/sync_test.go [hunk 7]
```diff
@@ -58,3 +65,117 @@ func TestSyncPendingCharts(t *testing.T) {
...
+			read, err := ioutil.ReadFile(dstIndex)
+			if err != nil {
+				t.Fatal(err)
+			}
+			newContents := strings.Replace(string(read), "https://fake.chart.repo.com/testing", fmt.Sprintf("%s%s", sTester.GetURL(), "/charts"), -1)
```
Comment (jotadrilo) [inline pkg/syncer/sync_test.go:132]:
```suggestion
			newContents := strings.Replace(string(read), "TEST_PLACEHOLDER", fmt.Sprintf("%s%s", sTester.GetURL(), "/charts"), -1)
```

File: pkg/syncer/sync_test.go [hunk 8]
```diff
@@ -58,3 +65,117 @@ func TestSyncPendingCharts(t *testing.T) {
...
+			// Create source and target testers
+			sTester, sCleanup, err := core.NewClientV2Tester(t, tc.sourceRepo.GetRepo(), false, dstIndex)
+			if err != nil {
+				t.Fatal(err)
+			}
+			tTester, tCleanup, err := core.NewClientV2Tester(t, tc.targetRepo.GetRepo(), true, "")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer sCleanup()
```
Comment (jotadrilo) [inline pkg/syncer/sync_test.go:124]: `defer cleanup()` goes right after you get the `cleanup` function

File: pkg/client/chartmuseum/chartmuseumtester.go (hunk 1 above)
Comment (tompizmor) [reply pkg/client/chartmuseum/chartmuseumtester.go]: modifies this and the others to start with a verb

File: pkg/client/core/coretester.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	"testing"
+
+	"github.com/juju/errors"
+
+	"github.com/bitnami-labs/charts-syncer/api"
+	"github.com/bitnami-labs/charts-syncer/pkg/client/chartmuseum"
+)
+
+// ClientV2Tester defines the methods that a fake tester should implement
+type ClientV2Tester interface {
```
Comment (jotadrilo) [inline pkg/client/core/coretester.go:14]: shouldn't this file be called `clienttester`, then?

**Review by jotadrilo — APPROVED** [review]
LGTM!

File: pkg/client/core/coretester.go (hunk 9 above)
Comment (tompizmor) [reply pkg/client/core/coretester.go]: just renamed

**Pull request merged**